cmd/kube-apiserver/app/options/validation.go

```diff
@@ -39,6 +41,36 @@ func validateClusterIPFlags(options *ServerRunOptions) []error {
 		errs = append(errs, errors.New("specified --service-cluster-ip-range is too large"))
 	}
 
+	// Secondary IP validation
+	secondaryServiceClusterIPRangeUsed := (options.SecondaryServiceClusterIPRange.IP != nil)
+	if secondaryServiceClusterIPRangeUsed && !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		errs = append(errs, fmt.Errorf("--secondary-service-cluster-ip-range can only be used if %v feature is enabled", string(features.IPv6DualStack)))
+	}
+
+	// note: While the cluster might be dualstack (i.e. pods with multiple IPs), the user may choose
+	// to only ingress traffic within and into the cluster on one IP family only. this family is decided
+	// by the range set on --service-cluster-ip-range. If/when the user decides to use dual stack services
+	// the Secondary* must be of different IPFamily than --service-cluster-ip-range
+	if secondaryServiceClusterIPRangeUsed {
+		// Should be dualstack IPFamily(ServiceClusterIPRange) != IPFamily(SecondaryServiceClusterIPRange)
+		dualstack, err := netutils.IsDualStackCIDRs([]*net.IPNet{&options.ServiceClusterIPRange, &options.SecondaryServiceClusterIPRange})
+		if err != nil {
+			errs = append(errs, errors.New("error attempting to validate dualstack for --service-cluster-ip-range and --secondary-service-cluster-ip-range"))
+		}
+
+		if !dualstack {
+			errs = append(errs, errors.New("--service-cluster-ip-range and --secondary-service-cluster-ip-range must be of different IP family"))
+		}
+
+		// should be smallish sized cidr, this thing is kept in etcd
+		// bigger cidr (specially those offered by IPv6) will add no value
+		// significantly increase snapshotting time.
+		var ones, bits = options.SecondaryServiceClusterIPRange.Mask.Size()
+		if bits-ones > 20 {
```

Comment:
should we document this constant? why 20?

Reply:
The subtle reason for this is the snapshots - save to etcd - executed frequently and maintained by the thing that assigns service IP. if we add more bits the snapshotting latency will increase. Same reason having too many service cidrs is going to also add lots of latency during snapshotting..

Reply:
@khenidak that's up to the admin though, how many service cidrs or size they want to use. I think we should document what the reasonable limits are, and maybe warn if they are exceeded?

Reply:
I think I'm even more in favor of having a single option after seeing all the shouldWorkOnSecondary() stuff and the split secondary stuff in the allocator below.